src/eval/Eval.ml

```diff
@@ -524,7 +522,7 @@ let init_contract clibs elibs cconstraint' cparams' cfields args' init_bal =
           tryM
             ~f:(fun (ps, pt) ->
               let%bind at = fromR @@ literal_type (snd a) in
-              if get_id ps = fst a && type_equiv pt at then pure true
+              if String.(get_id ps = fst a) && [%equal: typ] pt at then pure true
```

Comment:
Where is `[%equal: typ]` defined? Is it just the auto-generated equality function on the `typ` type, and this is then the way to ensure that that is the equality function being used?

Reply:
I found the answer in the next commit, so I'm resolving this one.

Reply:
`[%equal: foo]` refers to `equal_foo`, it need not be autogenerated. In particular `equal_typ` was introduced by Vaivas in Syntax.ml around the line 548 (and `typ` does not use `[@@deriving equal]`).